src/x86/sse.asm

```diff
@@ -0,0 +1,510 @@
+    paddd               m1, m2
+    movd                m2, [srcq+src_strideq*2]
+    movd                m3, [dstq+dst_strideq*2]
+    punpcklbw           m2, m3
+    movd                m3, [srcq+src_stride3q]
+    movd                m4, [dstq+dst_stride3q]
+    punpcklbw           m3, m4
+    pmaddubsw           m2, m0
+    pmaddubsw           m3, m0
+    pmaddwd             m2, m2
+    pmaddwd             m3, m3
+    paddd               m2, m3
+    paddd               m1, m2
+
+    pshuflw             m0, m1, q3232
+    paddd               m0, m1
+    movd               eax, m0
+
+    ; Multiply and shift using scalar code.
+    SSE_SCALE_4X4
+    RET
+
+%macro WEIGHTED_SSE_4X8_KERNEL 0
+    movd                m1, [srcq]
+    movd                m2, [srcq+src_strideq*4]
+    punpckldq           m1, m2
+    movd                m2, [dstq]
+    movd                m3, [dstq+dst_strideq*4]
+    add               srcq, src_strideq
+    add               dstq, dst_strideq
+    punpckldq           m2, m3
+    punpcklbw           m1, m2
+    movd                m2, [srcq]
+    movd                m3, [srcq+src_strideq*4]
+    punpckldq           m2, m3
+    movd                m3, [dstq]
+    movd                m4, [dstq+dst_strideq*4]
+    add               srcq, src_strideq
+    add               dstq, dst_strideq
+    punpckldq           m3, m4
+    punpcklbw           m2, m3
+    pmaddubsw           m1, m0
+    pmaddubsw           m2, m0
+    pmaddwd             m1, m1
+    pmaddwd             m2, m2
+    paddd               m1, m2
+    movd                m2, [srcq]
+    movd                m3, [srcq+src_strideq*4]
+    punpckldq           m2, m3
+    movd                m3, [dstq]
+    movd                m4, [dstq+dst_strideq*4]
+    add               srcq, src_strideq
+    add               dstq, dst_strideq
+    punpckldq           m3, m4
+    punpcklbw           m2, m3
+    movd                m3, [srcq]
+    movd                m4, [srcq+src_strideq*4]
+    punpckldq           m3, m4
+    movd                m4, [dstq]
+    movd                m5, [dstq+dst_strideq*4]
+    punpckldq           m4, m5
+    punpcklbw           m3, m4
+    pmaddubsw           m2, m0
+    pmaddubsw           m3, m0
+    pmaddwd             m2, m2
+    pmaddwd             m3, m3
+    paddd               m2, m3
+    paddd               m1, m2
+
+    %define LOAD_SCALES LOAD_SCALES_4X8
+    SSE_SCALE 1, 2, 3
+%endmacro
+
+INIT_XMM ssse3
+cglobal weighted_sse_4x8, 6, 6, 6, \
+        src, src_stride, dst, dst_stride, scale, scale_stride
+    mova                m0, [addsub]
+    WEIGHTED_SSE_4X8_KERNEL
+
+    pshufd              m0, m1, q3232
+    paddq               m1, m0
+    movq               rax, m1
+    RET
+
+INIT_XMM ssse3
+cglobal weighted_sse_4x16, 6, 7, 7, \
+        src, src_stride, dst, dst_stride, scale, scale_stride, \
+        h
+    mova                m0, [addsub]
+    pxor                m6, m6
+    mov                 hd, 1
```

Comment:
Just get rid of the loop here. You save a register and a (maybe well) predicted branch.
You can also just macro what's inside the loop and call it twice, would be the same number of lines.

Reply:
Done.